### PR TITLE
Add `Boto3ArtifactStore`

### PR DESCRIPTION
**Pull request by Alnusjaponica**
## Motivation
Migrate `Boto3ArtifactStore` from `optuna-dashboard` to `optuna`.

## Description of the changes
Move `Boto3Backend`, related dependency and tests from `optuna-dashboar` to `optuna`

### CODE REVIEW

File: optuna/artifacts/_filesystem.py
Comment (c-bata) [inline optuna/artifacts/_filesystem.py:23]: Just for the record, here we removed an example from the docstring since this example uses optuna-dashboard APIs. As a follow-up, we should add an example again after merging #4823.

File: optuna/artifacts/_boto3.py
Comment (c-bata) [inline optuna/artifacts/_boto3.py:36]: [nit] I've received feedback from Optuna Dashboard users suggesting that it's unclear which artifact wasn't found. It might be out of the scope of this PR, but what do you think about including the artifact_id in the error message?

```suggestion
                raise ArtifactNotFound(f"artifact={artifact_id} is not found") from e
```

File: optuna/artifacts/_boto3.py
Comment (c-bata) [inline optuna/artifacts/_boto3.py:56]: ditto.

```suggestion
                raise ArtifactNotFound(f"artifact={artifact_id} is not found") from e
```

File: optuna/artifacts/_boto3.py
Comment (c-bata) [inline optuna/artifacts/_boto3.py:68]:
```suggestion
    # A mypy-runtime assertion to ensure that Boto3ArtifactStore implements all abstract
    # methods in ArtifactStore.
```

File: tests/artifacts_tests/test_boto3.py
Comment (c-bata) [inline tests/artifacts_tests/test_boto3.py:14]: What do you think about writing these tests in a function-based manner since Optuna uses pytest runner?

**Review by c-bata — COMMENTED** [review]
Thank you for your pull request. Looks almost good to me, but I left some suggestions.

File: optuna/artifacts/_filesystem.py
Comment (Alnusjaponica) [reply optuna/artifacts/_filesystem.py:23]: Thanks for the comment. I opened https://github.com/optuna/optuna/issues/4843 to track this.

File: optuna/artifacts/_boto3.py
Comment (Alnusjaponica) [reply optuna/artifacts/_boto3.py:36]: Fair enough. Is it better to add bucket name as well?
```suggestion
                raise ArtifactNotFound(f"bucket: {self.bucket}, artifact: {artifact_id} was not found") from e
```

Comment (c-bata) [reply optuna/artifacts/_boto3.py:36]: Sounds great to me 👍

**Comment by toshihikoyanase** [comment]
@HideakiImamura Could you review this PR, please?

File: tests/artifacts_tests/test_boto3.py
Comment (Alnusjaponica) [reply tests/artifacts_tests/test_boto3.py]: I reformat them.

File: optuna/artifacts/_boto3.py
Comment (Alnusjaponica) [reply optuna/artifacts/_boto3.py]: I added some more explanation. Please take another look to confirm if they sounds natural.

**Review by c-bata — APPROVED** [review]
LGTM after passing CI.

File: optuna/artifacts/_boto3.py
Comment (HideakiImamura) [inline optuna/artifacts/_boto3.py:20]: How about adding the explanations for arguments?

File: optuna/artifacts/_filesystem.py
Comment (HideakiImamura) [inline optuna/artifacts/_filesystem.py:15]: ditto.

File: tests/artifacts_tests/test_boto3.py
Comment (HideakiImamura) [inline tests/artifacts_tests/test_boto3.py:84]: How about testing the case for `remove`?

File: tests/artifacts_tests/test_boto3.py
Comment (HideakiImamura) [inline tests/artifacts_tests/test_boto3.py:43]: How about adding the test case for `avoid_buf_copy` using `pytest.mark.parametrize`?

Comment (Alnusjaponica) [reply tests/artifacts_tests/test_boto3.py:44]: I added test cases.

File: tests/artifacts_tests/test_boto3.py
Comment (Alnusjaponica) [reply tests/artifacts_tests/test_boto3.py:88]: I added test to check `remove`

File: optuna/artifacts/_boto3.py
Comment (Alnusjaponica) [reply optuna/artifacts/_boto3.py]: Thanks for the comments. I added more explanations.

File: optuna/artifacts/_filesystem.py
Comment (Alnusjaponica) [reply optuna/artifacts/_filesystem.py:15]: ditto

File: optuna/artifacts/_boto3.py
Comment (Alnusjaponica) [inline optuna/artifacts/_boto3.py:77]: @c-bata 
It seems `client.delete_object` does not raise error when nothing found (See [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/delete_object.html#:~:text=Removes%20the%20null%20version%20(if%20there%20is%20one)%20of%20an%20object%20and%20inserts%20a%20delete%20marker%2C%20which%20becomes%20the%20latest%20version%20of%20the%20object.%20If%20there%20isn%E2%80%99t%20a%20null%20version%2C%20Amazon%20S3%20does%20not%20remove%20any%20objects%20but%20will%20still%20respond%20that%20the%20command%20was%20successful.)). Shall I just simply refactor like this?
```suggestion
        self.client.delete_object(Bucket=self.bucket, Key=artifact_id)
```

Comment (c-bata) [reply optuna/artifacts/_boto3.py:77]: @Alnusjaponica Thank you for the delayed response. Good catch. It sounds good 👍

**Comment by HideakiImamura** [comment]
Ops, could you try to resolve CI errors?

File: optuna/artifacts/_boto3.py
Comment (c-bata) [reply optuna/artifacts/_boto3.py]: @Alnusjaponica It's just an implementation detail that `Boto3Backend.remove()` does not raise an `ArtifactNotFound` exception, so there is no need to make it consistent with other ArtifactStores. If the `ArtifactNotFound` exception is not beneficial for the users, we might need to consider revising the following specification, but let's address it in a follow-up PR.

https://github.com/optuna/optuna/blob/master/optuna/artifacts/_protocol.py#L49-L59

Comment (Alnusjaponica) [reply optuna/artifacts/_boto3.py]: I see. Thanks for the comments. I mark this conversation as resesolved.

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Pull request merged**
